File: knowledge/dropbox/codex_stuff/ocr_vision.m
#import <Foundation/Foundation.h>
#import <AppKit/AppKit.h>
#import <Vision/Vision.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        if (argc < 3) {
            fprintf(stderr, "usage: ocr_vision OUTPUT.txt IMAGE...\n");
            return 2;
        }

        NSMutableString *finalText = [NSMutableString string];

        for (int i = 2; i < argc; i++) {
            NSString *path = [NSString stringWithUTF8String:argv[i]];
            NSURL *url = [NSURL fileURLWithPath:path];
            [finalText appendFormat:@"\n\n===== %@ =====\n", [url lastPathComponent]];

            VNRecognizeTextRequest *request = [[VNRecognizeTextRequest alloc] init];
            request.recognitionLevel = VNRequestTextRecognitionLevelAccurate;
            request.usesLanguageCorrection = YES;
            request.recognitionLanguages = @[@"en-US"];

            VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithURL:url options:@{}];
            NSError *error = nil;
            BOOL ok = [handler performRequests:@[request] error:&error];
            if (!ok) {
                [finalText appendFormat:@"[ocr failed: %@]\n", error ?: @"unknown"];
                continue;
            }

            for (VNRecognizedTextObservation *observation in request.results) {
                VNRecognizedText *candidate = [[observation topCandidates:1] firstObject];
                if (candidate != nil) {
                    [finalText appendFormat:@"%@\n", candidate.string];
                }
            }
        }

        NSString *outPath = [NSString stringWithUTF8String:argv[1]];
        NSError *writeError = nil;
        BOOL wrote = [finalText writeToFile:outPath atomically:YES encoding:NSUTF8StringEncoding error:&writeError];
        if (!wrote) {
            fprintf(stderr, "write failed: %s\n", [[writeError localizedDescription] UTF8String]);
            return 1;
        }
    }
    return 0;
}

File: knowledge/dropbox/codex_stuff/ocr_vision.swift
import Foundation
import Vision
import AppKit

let args = CommandLine.arguments.dropFirst()
guard args.count >= 2 else {
    fputs("usage: ocr_vision.swift OUTPUT.txt IMAGE...\n", stderr)
    exit(2)
}

let output = String(args.first!)
let imagePaths = args.dropFirst().map(String.init)
var finalText = ""

for path in imagePaths {
    let url = URL(fileURLWithPath: path)
    guard let image = NSImage(contentsOf: url),
          let tiff = image.tiffRepresentation,
          let bitmap = NSBitmapImageRep(data: tiff),
          let cgImage = bitmap.cgImage else {
        finalText += "\n\n===== \(url.lastPathComponent) =====\n[image load failed]\n"
        continue
    }

    let request = VNRecognizeTextRequest()
    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true
    request.recognitionLanguages = ["en-US"]

    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
    do {
        try handler.perform([request])
        let observations = request.results ?? []
        let lines = observations.compactMap { $0.topCandidates(1).first?.string }
        finalText += "\n\n===== \(url.lastPathComponent) =====\n"
        finalText += lines.joined(separator: "\n")
    } catch {
        finalText += "\n\n===== \(url.lastPathComponent) =====\n[ocr failed: \(error)]\n"
    }
}

do {
    try finalText.write(toFile: output, atomically: true, encoding: .utf8)
} catch {
    fputs("write failed: \(error)\n", stderr)
    exit(1)
}
